### ISSUE TITLE
how to perform optimization for arbitrary params

### ISSUE THREAD

**Issue by abd-dib** [BachiLi/redner#5]
Hi, I would like to thank you for sharing this code. Great work btw!!

is it possible to use ReDner to optimize for external params. let me give an example to be more clear,

suppose i want to optimize  Spherical Harmonics coefficients to estimate illumination at a certain point of the scene. I have my target image of a white sphere ( used as probe) and my sphere 3d model.

 In this case, vertices colors are updated  at each iteration using the normals and SH coeffs, than we render the new vertices and calculate the loss based on the target image. In this situation, we dont want to setup any lighting in the redner scene neither a materials. so my question is it possible for ur physically based renderer to handle this case ( with some tweaks ?)

Thank you

**Comment by BachiLi** [comment]
Hi,

Good questions! At the moment, environment map is not implemented (but it is pretty high on the priority), so the situation you describe can be a bit difficult. After we have the environment map lighting, you can write some PyTorch code to convert the spherical harmonics coefficients to the environment. The renderer will backpropagate to the pixels on the environment map, and you should be able to optimize for SH coefficients of the illumination by backpropagating in PyTorch.

I am still not sure about per vertex color/SH coefficients. Gouraud shading is just not a great model from the realistic rendering perspective. We may support per-face BRDF parameter and/or face tessellation in the future though. Right now the trick is similar to what I mentioned above: you can convert the color or SH coefficients into BRDF parameters to a texture in PyTorch (and make sure it is differentiable), then map it onto the surface. 

Once I finished the environment map implementation I can write an example for this. Let me know if this is not clear enough.

**Comment by abd-dib** [comment]
Clear enough! Thank you for your response and for the reference, very interesting!!
We will be waiting for the env mapping to be added!

**Comment by BachiLi** [comment]
Hey,
Sorry for the delay. I wrote an example regarding this, jointly optimizing SH coefficients of an environment map and the material parameters of a sphere.
https://github.com/BachiLi/redner/blob/master/examples/joint_material_envmap_sh.py
Let me know if this is what you want ; )

**Comment by abd-dib** [comment]
Ah!!!!! how nice is the tutorial u shared!!! it is exactly what we wanted to do with Redner.  from initial SH guess, reconstruct irradiance env map and optimize for it! that is a great example and it is exactly what we are looking for!! thank you so much for this great effort!!
 By the way i was looking at the open issues,  and i saw that there is an issue created for compiling on windows, I am happy to contribute, so i will share  modifications i did to compile redner on windows 10 with VS 2015.

**Comment by BachiLi** [comment]
Thanks a lot for your patch!!

**Comment by BachiLi** [comment]
Closing this issue. Feel free to reopen if there's other questions.
